Jetpack Plugin - Product description: Add summary before checkout

kallehauge opened the pull request:
## Introduction
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

This PR aims to add a new step when pressing some of the "Upgrade" buttons on the Jetpack plugin dashboard.
This step is added because we wish to give a bit more context of what the product/plan is before the customer is directed to the checkout page, but for now, by adding it to the plugin itself we also know that it will be translated through the normal release process.

## Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->

1. First off, we've repurposed `[PHP] Jetpack::get_products_for_purchase()` and its corresponding `[JS] <QuerySiteProducts>` query component (the data was no longer being used anywhere)._**_
2. Added a new component named `<JetpackProductCard>` which will responsively display a product/plan depending on what attributes it's provided with.
3. Added `<ProductDescriptions>` + `<ProductDescription>` components which are `<Switch>` + `<Route>` for the new Product Description views that take advantage of `<JetpackProductCard>`.
4. Added new images (already compressed) used in the product displays.
5. Added new icons used in the product displays.
6.  Replaced relevant "At a Glance" upgrade button links to new product descriptions.

_**  In the future it would be awesome to move more of it to an API - perhaps an API that actively tells you what discounts you can achieve for certain products (like we do in checkout) and start using it across multiple places in the plugin where we display product prices._

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

* PT: p1HpG7-cY7-p2
* Comment about differenced from the original design: p1HpG7-cY7-p2#comment-49164

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

This PR introduces 3 new tracks:

* `jetpack_product_description_view`: When a product description is shown. This doesn't take potential bundles into account but simply says e.g.: "You viewed the backup description view".
* `jetpack_product_card_view`: An event that is triggered for each product card shown, so on a product description view that contains e.g. Backup and Security Bundle, two instances of this event would be triggered (backup + security).
* `jetpack_product_card_checkout_click`: Triggered when someone has clicked "Add/Checkout" button in the product description view.

This changes the flows of the "At a Glance" / "Site Stats" upgrade links since upgrade links now redirect to product descriptions before sending the customer to checkout. All existing funnels etc should still be functional after this change since banner events has remained the same.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

**At a Glance: flow test**
* Spin up a site using this Jetpack branch.
* Go to `/wp-admin/?page=jetpack#/dashboard` aka _"At a Glance"_.
* Here you will find 4 upgrade buttons (actually 5, but we've only focused on the 4 highlighted below).
* Each of the 4 links should redirect to a product description view that looks like the screenshots in the screenshot below.
* When you're on one of the new product description views, press the `Add Jetpack <product>` button and make sure you're take to Jetpack Checkout.

![Screenshot 2021-09-27 at 17 50 05](https://user-images.githubusercontent.com/3846700/134942342-588f6e78-9dd7-4878-80ff-c7d80be911bc.png)

**Site Stats: flow test**
* Spin up a site using this Jetpack branch.
* Go to `/wp-admin/admin.php?page=stats` aka _"Site Stats"_.
* Here you will find 4 upgrade buttons for backup, search, akismet & scan. 
* Each of the 4 links should redirect to a product description view that looks like the screenshots in the screenshot below.
* When you're on one of the new product description views, press the `Add Jetpack <product>` button and make sure you're take to Jetpack Checkout.

![Screenshot 2021-10-04 at 17 13 02](https://user-images.githubusercontent.com/3846700/135877606-63dde320-8abd-4a21-88ea-69c9b10996e8.png)

**Tracks**

If you've done the above testing, you should automatically have triggered all of the new tracks as well (documented in the bullets below for easy of testing):

* `jetpack_product_description_view`
* `jetpack_product_card_view`
* `jetpack_product_card_checkout_click`

When you've done the above, you should then have triggered all 3 tracking events mentioned in the description of this PR.
If you wait ~10min you should be able to validate that the data was registered in our live tracking tool (PCYsg-5hv-p2) and optionally filter by your WPCOM username - even though there will probably not be that many events, so just look for your own username to make sure it's the ones you triggered.

## Screenshots

### "Desktop"

![screenshot-2021-09-27-at-14 59 55](https://user-images.githubusercontent.com/3846700/134942871-9f39c7d9-281c-4f7c-8551-0f005cb66905.png)
![screenshot-2021-09-27-at-15 00 28](https://user-images.githubusercontent.com/3846700/134942885-1f172fba-8701-442b-baa1-f4217da49aa8.png)
![screenshot-2021-09-27-at-15 00 43](https://user-images.githubusercontent.com/3846700/134942888-38a89414-86a0-4cce-9d63-97074ddccaa4.png)
![screenshot-2021-09-27-at-14 59 29](https://user-images.githubusercontent.com/3846700/134942890-f9f5a5fb-2b13-476a-9e30-04124ed468a1.png)

### "Tablet"

![screenshot-2021-09-27-at-15 30 33](https://user-images.githubusercontent.com/3846700/134943038-2429f609-f795-4200-86a6-84e4767a3b81.png)
![screenshot-2021-09-27-at-15 30 22-1](https://user-images.githubusercontent.com/3846700/134943040-37e55d08-6488-4baa-a4fe-4f4e41db971e.png)

### "Phone"

![screenshot-2021-09-27-at-15 29 31-1](https://user-images.githubusercontent.com/3846700/134943074-b635c075-66d3-4bd3-8a6f-4b881a165a26.png)
![screenshot-2021-09-27-at-15 29 57-1](https://user-images.githubusercontent.com/3846700/134943076-5307b2de-f529-463a-85df-8f28ea1213f0.png)

kallehauge committed:
Jetpack Plugin: Generalise Recommendations Loading Card (JetpackLoadingIcon)

kallehauge committed:
Jetpack Plugin: Repurpose Site Products and update <QuerySiteProducts>

QuerySiteProducts would throw errors when it was used like most of the other data query components.

kallehauge committed:
Jetpack Plugin: Add new Jetpack product illustrations

kallehauge committed:
Jetpack Plugin: Add Product Description views

kallehauge committed:
Jetpack Plugin: Reduce size of Jetpack product illustrations

We found that the images where approximately where sometimes 50% larger than they had to be, after "finishing" the product description view.

kallehauge committed:
Add changelog

kallehauge committed:
Jetpack Plugin: Improve small screen styling of new product details views

kallehauge committed:
Jetpack Plugin: Remove <QuerySiteProducts> from lint exclude list

kallehauge committed:
Jetpack Plugin: Fix broken Recommendation Summary loading tests

kallehauge committed:
Update changelog message

kallehauge committed:
Unify constant code styles and remove commented out code

kallehauge committed:
Merge branch 'master' into add/jp-product-description-view

kallehauge committed:
Jetpack Plugin: Redirect from At a Glance to product descriptions

kallehauge committed:
Jetpack Plugin: Move <LoadingCard> to be generally accessible

kallehauge committed:
Merge branch 'update/refactor-loading-card' into add/jp-product-description-view

kallehauge commented:
Aside: I'm currently preparing tests for this code as well but it shouldn't change much in terms of functional/technical review 🙂

keoshi commented:
Hey @kallehauge — I couldn't test this on my JN site. The CTAs were using these URLs:
- `https://jetpack.com/redirect/?source=aag-backups&site=liable-snare.jurassic.ninja&u=1&query=`
- `https://jetpack.com/redirect/?source=aag-scan&site=liable-snare.jurassic.ninja&u=1&query=`

Not sure if I'm doing anything wrong, but will try again tomorrow on a different environment.

kallehauge commented:
@keoshi I got it to work by doing the following:

* Go to JN/specialops
* Select "Include Jetpack Beta"
* Select `add/jp-product-description-view` as the Jetpack Branch
* Make a Jetpack Free connection
* Go to "At a Glance" (`/wp-admin/admin.php?page=jetpack#/dashboard`)

It's probably also what you were doing but please let me know if it's still not working for you. Alternatively feel free to ping me on Slack and we'll see if it's the environment or the code that needs more attention :)

keoshi reviewed (changes requested):
Hey @kallehauge! I finally could test this today, not sure why it wasn't working before.

It all looks really good, thanks for all your hard work on this. Just noticed a couple of things:

- Only the links in At a Glance are affected. Just wanted to make sure we're also changing the  others we documented in the CTAs audit: p1HpG7-cTd-p2
- When the upsell is present, the content on the panes doesn't line up perfectly. It's barely a few pixels (5), but if we could fix this it would be 💯 

![image](https://user-images.githubusercontent.com/390760/135266124-a014c5d3-7915-49ab-aa06-7672b42b0171.png)

- In the screen for Scan, the icon on the left is much smaller than the one on the right:

![image](https://user-images.githubusercontent.com/390760/135266737-f117565f-247a-428f-8ac5-d00bff0cc738.png)

- In the screen for Search:
  - The title and CTA should read “Jetpack Site Search”.
  - There's no information on the current number records and tier upper limit.
  - The image is not retina-ready. We should use an image at 2x the resolution or 1.5x at least.
- Finally, do we have tracks events for all these new flows?

Thank you!

kallehauge committed:
Merge branch 'master' into add/jp-product-description-view

kallehauge committed:
Jetpack Plugin: Add tracking to Jetpack Product Card checkout click

kallehauge committed:
Jetpack Plugin: Add Jetpack Product Card view event

kallehauge committed:
Jetpack Plugin: Add Jetpack Product Description view event

kallehauge committed:
Merge branch 'master' into update/refactor-loading-card

kallehauge committed:
Merge branch 'update/refactor-loading-card' into add/jp-product-description-view

kallehauge committed:
Jetpack Plugin: Generalize products for purchase upgrade URL

We do this so the data can be used outside of At a Glance.

kallehauge commented:
Reasoning for the label spam above: I merged master into this branch before doing the same to the `update/refactor-loading-card` branch which this is based upon, so it basically took all of the tags from the 10.2 release and applied them here 😓

kallehauge committed:
Jetpack Plugin: use product description in Site Stats upgrades

kallehauge committed:
Jetpack Plugin: Change Jetpack Search product title

kallehauge committed:
Jetpack Plugin: change Jetpack Search upgrade link on Site Stats

We wish to display a product description before sending the user to checkout as to provide more context before just asking for payment.

kallehauge committed:
Jetpack Plugin: Align product description primary and secondary cards

kallehauge commented:
Thank you for reviewing @keoshi ! 🎉 

> Only the links in At a Glance are affected.
Now also affects Site Stats!

> When the upsell is present, the content on the panes doesn't line up perfectly.

Should now be corrected. Nice catch! 👏 

> In the screen for Scan, the icon on the left is much smaller than the one on the right:
Again, nice catch! 👏  It should not have the exact same size as the bundle version of the shield.

> Search tier data
Oh, I completely overlooked this during implementation. I think this has to be implemented after/outside of this PR since it's already at a big size and adding more to it will just make it more difficult to get approved.
I'll make sure this is a follow up task!

* Finally, do we have tracks events for all these new flows?
Yes, I've updated the **Does this pull request change what data or activity we track or use?** section with tracks requested by marketing 😄

kallehauge committed:
Jetpack Plugin: Add JetpackProductCard tests

kallehauge committed:
Jetpack Plugin: Add ProductDescription tests

kallehauge committed:
Jetpack Plugin: Update JetpackProductCard tests with features

kallehauge committed:
Merge branch 'master' into update/refactor-loading-card

kallehauge committed:
Jetpack Plugin: Add altText property to JetpackLoadingIcon

kallehauge committed:
Remove unneccesary new lines in changelog file

kallehauge committed:
Merge branch 'update/refactor-loading-card' into add/jp-product-description-view

jeherve commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -7398,94 +7398,78 @@ public static function is_active_and_not_offline_mode() {
+			'description'       => __( 'Save time and get better responses by automatically blocking spam from your comments and forms.', 'jetpack' ),
+			'show_promotion'    => true,
+			'discount_percent'  => 40,
+			'included_in_plans' => array( 'security' ),
```
Shouldn't we leave the legacy plans in there?

kallehauge replied:
> Shouldn't we leave the legacy plans in there?

For the new use-case (product description "pages") we're, for now, only planning on upselling the security bundle and are not planning on upselling legacy bundles which means we don't need the data. We can still leave it there but we would leave it without any particular use-case which IMO can be confusing for other developers (e.g.: are there related legacy plan code elsewhere etc).

I'm up to both solutions, so if you feel strongly about one over the other, please let me know and I'll update if it necessary 😄

jeherve commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -7398,94 +7398,78 @@ public static function is_active_and_not_offline_mode() {
 	 */
 	public static function get_products_for_purchase() {
 		$products = array();
-		if ( ! is_multisite() ) {
```
Does this mean that backups can now work on Multisite networks, or should this remain?

kallehauge replied:
> Does this mean that backups can now work on Multisite networks, or should this remain?

Even though, technically, I think we now offer a download link instead of an actual restore (I think this is only for individual sites on a multisite installation and not the network), then no. Multisite Networks are not officially supported :)
The main reason I'm removing it here is because if someone links to Jetpack Backup's product description, then I still want to show it.

I few options if we want to change it back:

* We create a fallback page for "dead" product description links that gives a generic message like e.g.: "Product is not available on your site" if a multisite site ends up on e.g. the backup product description.
* We rename all references from "products for purchase" to "product descriptions" to avoid any confusions here if the listed products are available for the site itself.

I would like the product description page to be as "dumb" as possible and just show the description if the user ends up on the "page". Then it would be the place that implements the link that is responsible for if it should be shown or not.

Context: I basically took the "products for purchase" and repurposed it since it was completely deprecated anyway (unused, fatal error query component, ignored for lints etc) and a lot of the logic could be reused for a new use-case and in my mind it still made sense with "for purchase" since it's basically sales texts.

jeherve replied:
> I would like the product description page to be as "dumb" as possible and just show the description if the user ends up on the "page". Then it would be the place that implements the link that is responsible for if it should be shown or not.

That makes sense. Let's leave it as it is then, that seems okay as long as we're mindful of that.

> Then it would be the place that implements the link that is responsible for if it should be shown or not.

Do you think we could add a comment on the method mentioning that somehow? Would that be overkill?

jeherve commented on projects/plugins/jetpack/modules/stats/class-jetpack-stats-upgrade-nudges.php:
```diff
@@ -301,6 +301,16 @@ private static function get_upgrade_link( $source ) {
+	 * @return string
+	 */
+	private static function get_product_description_link( $product_key ) {
+		return admin_url( sprintf( '?page=jetpack#/product/%s', $product_key ) );
```
We have a `Jetpack::admin_url()` method that may be useful here.

kallehauge commented on projects/plugins/jetpack/_inc/client/components/jetpack-product-card/index.jsx:
```diff
@@ -0,0 +1,159 @@
+
+	return (
+		<div className={ classes }>
+			<sup className="jp-product-card__currency-symbol">{ priceObject.symbol }</sup>
+			<span className="jp-product-card__price-integer">{ priceObject.integer }</span>
+			<sup className="jp-product-card__price-fraction">{ priceObject.fraction }</sup>
+		</div>
```
I don't think it does because even though the passed value is `countryCode` (e.g.: `USD`), we actually display the symbol instead (e.g.: `$`).
So for danish currency it would say `kr` which is the same no matter the language (even though I personally prefer DKR to differentiate between e.g. Swedish Kr and Danish Kr 😛 ).

I've copied the logic from cloud.jetpack.com where they do it the same way without translating it (not saying this is a stamp of approval, but I don't know of any currency symbol issues over there).

kallehauge committed:
Jetpack Plugin: Update Product Description link Jetpack logic

kallehauge committed:
Jetpack Plugin: Correct Jetpack Scan product description link

kallehauge committed:
JetpackLoadingIcon: translate default alt text prop

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

keoshi commented:
Hey, @kallehauge! This is looking really good, only have two notes:

- I'm not redirected to the detailed view consistently. Sometimes I get the correct view, others I'm redirected to the checkout as before. I assume this is caused by JP Beta or the JN environment so I'm not worried.
- We've proposed a simplification for the upsell section in the i1 designs thread p1HpG7-cTd-p2, so we wouldn't necessarily need the security bundle illustration/copy. Would look something like this:

![image](https://user-images.githubusercontent.com/390760/136232912-a4d3c2c2-bd90-4949-a1c4-d2f0a7fc8ccb.png)

kallehauge committed:
Jetpack Plugin: Use ellipsis in JetpackLoadingIcon default prop

kallehauge committed:
Merge branch 'master' into update/refactor-loading-card

jeherve committed:
Merge remote-tracking branch 'origin/master' into update/refactor-loading-card

kallehauge committed:
Merge branch 'master' into update/refactor-loading-card

kallehauge committed:
Merge branch 'update/refactor-loading-card' into add/jp-product-description-view

kallehauge committed:
Merge branch 'master' into add/jp-product-description-view

kallehauge committed:
Jetpack Plugin: Include "admin.php" in getProductDescriptionUrl

kallehauge committed:
Merge branch 'master' into add/jp-product-description-view

kallehauge committed:
Jetpack Plugin: Refactor QuerySiteProducts to a function component

kallehauge committed:
Jetpack Plugin: Hide JetpackProductCard features <ul> when empty

kallehauge committed:
Jetpack Plugin: Use @automattic/color-studio colors for <JetpackProductCard>

kallehauge committed:
Jetpack Plugin: Add description to Jetpack::get_products_for_purchase

kallehauge committed:
Jetpack Plugin: Upload larger sizes of product illustrations

kallehauge commented:
@keoshi some responses  😄 

> Oh, just noticed something else: when the detailed view is loaded, the side navigation incorrectly points to the WP Dashboard, not Jetpack:

Corrected! The Jetpack Dashboard is now shown instead 😅 

> I'm not redirected to the detailed view consistently. Sometimes I get the correct view, others I'm redirected to the checkout as before. I assume this is caused by JP Beta or the JN environment so I'm not worried.

I'm not that worried about it either but please let me know if your level of concerns increases 😝 

> We've proposed a simplification for the upsell section in the i1 designs thread p1HpG7-cTd-p2, so we wouldn't necessarily need the security bundle illustration/copy.

As a fallback (or at least for now), I'll continue to use the Scan illustration for Security as well - just so it has an associated image. And it looks like some security stuff with the shield and everything 😄

kallehauge committed:
Jetpack Plugin: discountedPrice on JetpackProductCard is, now, always a number

kallehauge commented:
> Sounds good. We don't have a Security illustration yet, but I'll take a look at it tomorrow and see if anything good comes out of it. :)

Sounds great @keoshi!

In terms of the PR itself, I don't feel like the illustration should hold us back from merging this @IanRamosC / @jeherve, if you don't have anything else.
There are upcoming work depending on this PR, so potentially changing the illustration in a separate/new PR shouldn't block us 😄

kallehauge committed:
Merge branch 'master' into add/jp-product-description-view

keoshi reviewed (dismissed):
LGTM! Thanks for all your work on this, @kallehauge.

IanRamosC reviewed (dismissed):
Retested it and it looks good. Nice improvements!

jeherve reviewed (changes requested):
This tests well for me. I have a few comments, below:

-------

Not a blocker, I think this can be added in a follow-up PR: we're missing the footer at the bottom of the product screens:
<img width="1242" alt="Screenshot 2021-10-15 at 17 34 10" src="https://user-images.githubusercontent.com/426388/137514620-977f73ce-b07d-4fb3-b1bc-59b282eb6136.png">

You can find that footer in other pages of the dashboard.

-------

In a follow-up PR, could we add the same screen for the VideoPress upgrade CTA, for consistency?

kallehauge committed:
Jetpack Plugin: Site Stats - link search nudge to product description

kallehauge committed:
Jetpack Plugin: Show Jetpack footer on product description pages

kallehauge committed:
Jetpack Plugin: Description - reduce spacing between content and footer

We do this so it matches with the rest of the JP admin pages.

kallehauge committed:
Merge branch 'master' into add/jp-product-description-view
